hail/src/main/scala/is/hail/backend/service/Worker.scala

```diff
@@ -44,21 +69,114 @@ class WorkerTimer() {
+    val original = Thread.currentThread().getContextClassLoader()
+    try {
+      Thread.currentThread().setContextClassLoader(custom)
+      f
+    } finally {
+      Thread.currentThread().setContextClassLoader(original)
+    }
+  }
+
+  private[this] val STRING_ARRAY_CLASS = Class.forName("[Ljava.lang.String;")
+  private[this] def loadMainFromLocalJar(revision: String, localJarURL: URL): Array[String] => Unit = {
+    val classLoader = new LoadSelfFirstURLClassLoader(Array(localJarURL))
+    val reflectedMain = withThreadContextClassLoader(classLoader) {
+      val workerClass = classLoader.loadClass("is.hail.backend.service.Worker")
+
+      val actualRevision = workerClass.getMethod("myRevision").invoke(null).asInstanceOf[String]
+      assert(revision == actualRevision, s"$revision != $actualRevision")
+
+      workerClass.getMethod("main", STRING_ARRAY_CLASS)
+    }
+
+    { (args: Array[String]) =>
+      withThreadContextClassLoader(classLoader) {
+        try {
+          reflectedMain.invoke(null, args): Unit
+        } catch {
+          case e: InvocationTargetException =>
+            throw e.getCause()
+        }
+      }
+    }
+  }
+
+  private[this] def downloadFile(remoteFS: FS, remote: String, local: String): Unit =
+    retryTransientErrors {
+      using(remoteFS.openNoCompression(remote)) { is =>
+        using(new FileOutputStream(local)) { os =>
+          IOUtils.copy(is, os)
+        }
+      }
+    }
+
+  private[this] def mainForRevision(revision: String, remoteJarLocation: String): Array[String] => Unit = {
+    revisions.get(revision) match {
+      case Some(mainMethod) =>
+        log.info(s"$revision found in the memory cache")
+        mainMethod
+      case None =>
+        log.info(s"$revision not in the memory cache")
+
+        val localJarLocation = s"/hail-jars/${revision}.jar"
+        val localJarFile = new File(localJarLocation)
+        if (!localJarFile.exists()) {
+          log.info(s"$revision not in the disk cache")
+          localJarFile.getParentFile().mkdirs()
+          downloadFile(workerGCSFS, remoteJarLocation, localJarLocation)
+          log.info(s"$revision added to the disk cache at $localJarLocation")
+        }
+
+        val localJarURL = localJarFile.toURI().toURL()
+        val mainMethod = loadMainFromLocalJar(revision, localJarURL)
+        log.info(s"$revision loaded")
+        revisions.put(revision, mainMethod)
+        mainMethod
     }
-    val root = args(0)
-    val i = args(1).toInt
+  }
+
+  def main(args: Array[String]): Unit = {
     val timer = new WorkerTimer()
+    timer.start("main")
 
-    var scratchDir = System.getenv("HAIL_WORKER_SCRATCH_DIR")
-    if (scratchDir == null)
-      scratchDir = ""
+    log.info(s"is.hail.backend.service.Worker $myRevision")
+    if (args.length < 2) {
+      throw new IllegalArgumentException(s"expected at least two arguments, not: ${ args.length }")
+    }
```

Comment:
I feel like you might as well check for 4 arguments here instead of checking once for 2 and then for 4.

Reply:
I considered this, but I want to allow us to change the number of arguments without breaking backwards compatibility.